Added HW volume control for TAS57xx

flrahe opened the pull request:
This is a propose for implementing the HW volume control of the TAS57xx on the squeezelite-ESP32.

This is my first pull request. Hope I'm doing it right.
Regards,
Florian

sle118 commented:
Just a small comment here. We typically try to exclude sdkconfig from commits to keep it as neutral as possible unless introducing new configuration variables

philippe44 commented:
Thanks - it looks good globally. Can you remove sdkconfig from the PR? Can you also address if/else style consistency? We all have our preferences, but I think regardless of these, it is important to "merge" with the style of a file one makes modification into. When you're are the creator, you can use yours :-)

flrahe commented:
My fault. Ist's Just due to my missing knowledge about git.
I'll try to look for that the next time.
Florian

Sébastien <notifications@github.com> schrieb am So., 6. Sept. 2020, 22:46:

> Just a small comment here. We typically try to exclude sdkconfig from
> commits to keep it as neutral as possible unless introducing new
> configuration variables
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/sle118/squeezelite-esp32/pull/51#issuecomment-687897154>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOIBWYGJDQ5QR2X6ZORPVQ3SEPYIXANCNFSM4Q4ZTKZQ>
> .
>

flrahe commented:
Cleaned up code and reduced some outputs to log.
Reverted sdkconfig to original.
Used if/else formatting for style consistency.

There are some additional changes in the code. With the equalizer we came into clipping at around +10dB. It seems to me a result of FIXED_ONE software gain. Reducing it to FIXED_ONE>>1 = FIXED_ONE/2 (approx 87% volume setting) and increasing hardware gain of the amplifier accordingly results in better behavior (limit around +14dB). We loose here 1 bit of resolution, but get space for the equalizer. Seems to be better than loosing approx. 4 bits when working with a volume of 50% (would be FIXED_ONE>>4) which seems to be a typical operation setting of this amplifier.

philippe44 commented on components/squeezelite/tas57xx/dac_57xx.c:
```diff
@@ -130,9 +139,68 @@ static void deinit(void)	{
+	i2c_master_write_byte(i2c_cmd, (uint8_t) left, I2C_MASTER_NACK);
+	ESP_LOGI(TAG, "i2c write %x at %u", 0x3d, (uint8_t) left);
+
+	if (left > 0) {
```
Are you sure about that?

flrahe replied:
Absolutely not! Very good eyes and review. It was a copy / paste error. Removed.

sle118 commented:
@philippe44 are we ready for a squash and merge?

philippe44 commented:
I need to review the impact of not setting FIXED_ONE which has a special meaning

flrahe commented:
> I need to review the impact of not setting FIXED_ONE which has a special meaning

I can understand you. My first solution was also setting FIXED_ONE. Then I came into clipping when applying equalizer values above +11dB at low frequencies. Since the equalizer seems to be a wildly used option (when using smaller speaker), I thought of improving that.

flrahe commented:
> I need to review the impact of not setting FIXED_ONE which has a special meaning

Any news on that? I had that setup now running for some time and it's working perfectly. Yesterday I patched the actual cmake-version to be on the newest release. It seems to me that sound is more clearer and precise compared to SW-volume, but it's just a personal impression.
Maybe we can discuss on my proposal?

sle118 commented:
I believe @philippe44 is currently busy with "reality" and can't dedicate a lot of time on the project. I will try to review the PR myself and possibly make it a configurable option that can be switched off in case it causes problems in some other usecases.

philippe44 commented:
I'll do that during the week-end

philippe44 commented:
So I had a look. Are you sure this changes something? The TAS volume is digital as well, it's not an analogue control of any kind (which probably is funny with a class-D). So there is no difference in making it in esp32 or in TAS. It's 16 bits-multiply on a 32 bits processor, there is no more or less loss

flrahe commented:
I see your point. In the TAS there is only little description about the volume control. It is only clear, that the amplifier gain won't be changed. So the calculation will be done somewhere in the DSP side of the TAS. It's the question whether the calculation in the DSP makes any difference from yours. Since there isn't any documentation, no one can really tell. Only thing would be to test by hearing, but to be honest, I suppose the result will depend on the listener...

The part with FIXED_ONE wasn't obvious to me. Therefore I see that my choice is not good from that point of view.

To my personal preference, I'd like to see a volume control by the TAS realized as an option, possibly with going back to FIXED_ONE to overcome the limitations for VU/spectrum. It's up to you, since there is no obvious technical point for that.

philippe44 commented:
I don't know how the gain is applied in the TAS, but it should have no compression/non-linearity so a plain multiplication will be same wherever it happens. 

I would agree to let the chip do the volume if it is an analogue (not entering debate if this is really better from and SNR point of view) but here it would just force me to do a special handling of VU and Spectrum when user wants to use SW gain in the TAS which means propagating information that the display does not fully have for now. It has knowledge of gain to compared it to FIXED_ONE and be able to scale differently then, but it would when then another information that give the potential gain although it's not applied. 

I would prefer to keep this as a potential improvement when we'll have really the case of device with analogue gain rather that adding another NVS parameter. Too many choices are not a good choice

sle118 commented:
We need to make a decision here. Either we don't implement the change, or we make it optional with an NVS parameter that I could expose through the UI. In any cases this should be resolved. Thank you!

philippe44 commented:
We can leave it open and tag it for optional consideration, later. I don't think this is needed as of today

flrahe commented:
Sorry for answering so late.
We may also close this request and you please keep this solution in mind. I understand all the implications and the effects on other parts of the code.

If one wants a nice stylish player, all these things are needed.

Mine is working headless in a dark corner, but that's not the case for others.

My hope is that you'll keep the code slim and clean. With too many features it'll may get more unstable. I for myself will change my code back to FIXED_ONE and will patch your code from time to time to update my player (have now build four of them). Now I know how to compile the code and that's working well. Now I need to learn how to do that patching automatically... :-)

Greetings Florian
